In [0]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, avg, when, count, floor, datediff, min, max
from pyspark.sql import functions as F
from pyspark.sql.window import Window

In [0]:
# load related dataset from S3
df_pitstops = spark.read.csv('s3://columbia-gr5069-main/raw/pit_stops.csv', header=True)
df_results = spark.read.csv('s3://columbia-gr5069-main/raw/results.csv', header=True)
df_drivers = spark.read.csv('s3://columbia-gr5069-main/raw/drivers.csv', header=True)
df_races = spark.read.csv('s3://columbia-gr5069-main/raw/races.csv', header=True)
df_laptimes = spark.read.csv('s3://columbia-gr5069-main/raw/lap_times.csv', header=True)
df_sprint_results = spark.read.csv('s3://columbia-gr5069-main/raw/sprint_results.csv', header=True)


1. [10 pts] What was the average time each driver spent at the pit stop for each race?


In [0]:
avg_pit_stop = df_pitstops.groupBy("raceId", "driverId").agg(avg("milliseconds").alias("avg_pit_time"))
display(avg_pit_stop)

raceId driverId avg_pit_time 843 39 26049.0 844 67 21976.333333333332 844 2 21743.75 844 20 20402.25 849 17 24416.0 856 20 20077.5 869 819 18606.0 873 818 30198.5 876 20 20707.0 878 3 21346.0 880 154 22302.0 887 8 24710.0 909 18 20247.0 911 822 22972.5 957 837 30066.5 958 822 21554.5 970 8 20569.5 986 1 26299.0 1009 844 21828.0 1009 154 22354.0 1022 8 30314.0 1035 817 28270.333333333332 1043 826 30043.0 1047 849 21870.5 1059 4 31288.0 1061 844 1038082.0 1065 839 30333.0 1067 20 23507.0 1067 840 35509.0 1093 830 28060.0 1051 832 25639.0 843 808 21993.0 844 17 21100.25 860 3 23110.0 861 30 23696.0 866 813 22704.0 880 820 30116.666666666668 882 824 22185.333333333332 884 821 20384.5 891 4 22537.5 894 813 23556.0 895 820 24010.5 896 822 26764.0 897 813 23148.0 899 819 24099.0 908 817 28483.0 941 822 30316.0 942 4 27341.0 955 831 20691.5 969 832 21762.0 982 838 29834.8 982 832 28897.5 996 828 24295.0 1000 828 22640.0 1037 825 25247.5 1047 8 22012.0 1060 832 21431.0 1061 852 1052414.5 1072 853 1019226.0 1073 1 21595.0 1078 842 19275.0 1079 4 23232.0 1079 854 23354.5 1084 4 22543.666666666668 1098 846 32555.666666666668 1110 839 23212.0 985 817 23604.0 1104 825 51052.0 847 22 24028.5 848 10 21965.0 863 20 22178.0 871 39 23238.0 931 154 25077.0 948 1 556013.5 950 3 23195.0 962 835 27620.666666666668 978 815 28942.0 979 815 24017.0 1012 807 24107.0 1020 8 22908.75 1035 825 29161.5 1044 20 24711.5 1052 822 27241.666666666668 1058 1 21429.5 1074 825 25801.666666666668 1094 830 22487.0 1110 807 24904.5 895 818 24926.333333333332 975 815 23229.0 1061 849 1052353.5 880 821 22315.5 904 13 22525.0 912 813 24223.0 913 824 30016.666666666668 934 828 30122.5 1013 20 19981.0 1020 840 21308.4 1022 20 22824.5 1037 1 23298.0 1039 1 560077.6 1053 842 393241.75 1060 847 21606.0 1051 852 26166.5 1072 1 701036.6666666666 1078 1 18820.0 1088 1 17455.0 1093 842 26294.0 1098 856 25126.0 1100 822 610372.6666666666 991 828 22785.0 844 813 20547.0 858 30 20149.0 873 807 31859.0 876 39 24544.0 892 821 24668.0 906 3 24492.0 909 8 19781.333333333332 943 826 22538.5 972 836 31032.5 980 822 22941.0 998 838 28471.0 1004 4 30348.0 1031 842 21188.5 1087 855 23554.0 1096 20 23186.0 1100 856 438992.75 895 807 23607.0 986 815 22715.5 995 807 24521.0 1101 842 20702.0 847 2 23238.5 851 808 23172.0 901 821 27872.5 901 3 24888.0 902 18 24860.5 918 20 22352.5 929 826 25056.5 939 13 28567.333333333332 973 828 22885.0 979 1 21338.0 983 839 24456.0 995 842 24107.0 997 825 22138.0 1016 848 37950.0 1022 826 22834.0 1031 847 21417.0 1095 847 23500.5 1102 4 22631.0 878 808 22762.0 841 5 24865.0 847 39 29327.5 884 822 20537.333333333332 902 822 25362.666666666668 939 829 34122.5 949 828 25069.0 950 807 25977.5 959 831 23188.0 961 832 24821.0 963 817 24302.5 968 828 22991.0 998 1 28982.0 999 838 19699.5 1021 826 23002.0 1029 832 22770.0 1031 822 19172.0 1032 1 21729.0 1043 840 34461.666666666664 1053 20 321415.2 1061 854 1052523.5 1062 817 504414.0 1078 825 20216.5 1083 844 1532582.5 1085 855 43617.0 1087 848 23874.5 1101 846 20346.0 907 815 22865.0 864 815 23816.666666666668 868 817 25619.5 873 808 34512.0 901 18 24797.666666666668 915 4 35112.0 930 13 22341.666666666668 943 831 23418.666666666668 945 3 21608.5 956 4 19940.25 960 13 348554.3333333333 966 826 22765.0 980 828 23670.666666666668 981 830 32030.5 988 836 22959.0 993 20 23163.5 1019 846 28097.0 1035 1 28904.0 1070 853 24365.5 1083 4 1037758.0 1089 846 25749.0 1092 839 209386.5 1102 1 22150.0 1109 847 21780.5 1047 825 22906.0 855 13 21209.0 874 30 20511.0 887 154 25129.0 888 1 19374.666666666668 906 8 24073.0 936 826 19922.2 943 822 23058.5 943 828 23236.333333333332 957 1 29535.5 964 154 24585.0 966 825 22327.5 973 832 22183.0 973 154 22753.5 983 840 24109.0 1017 844 30203.0 1040 847 29757.666666666668 1042 1 27162.0 1067 815 22463.0 1082 4 24459.0 1106 852 25584.5 1108 4 28679.0 869 808 18826.333333333332 985 828 25841.0 855 20 21891.666666666668 855 67 21849.0 860 808 24579.0 865 819 27186.0 875 17 19713.5 886 1 20767.5

2. [20 pts] Rank the average time spent at the pit stop in order of who won each race

In [0]:
# How you decided to deal with drivers who did not finish the race: 
# Drivers who do not finish the race typically have a NULL value in the position column of the results dataset.
# I chose to exclude all drivers with NULL position by using isNotNull:
# Filter out drivers who did not finish (no position)
results_filtered = df_results.filter(col("position").isNotNull()) \
    .select("raceId", "driverId", "position")
# Join with average pit stop times
joined_df = avg_pit_stop.join(results_filtered, on=["raceId", "driverId"], how="inner")
# Add driver name and race name from other datasets
driver_names = df_drivers.select("driverId", "forename", "surname")
race_names = df_races.select("raceId", "name")
# Select columns and order by position joined result
ranked_avg_pitstops = joined_df \
    .join(driver_names, "driverId", "left") \
    .join(race_names, "raceId", "left") \
    .select("raceId", "name", "driverId", "forename", "surname", "position", "avg_pit_time") \
    .orderBy("raceId", F.col("position").cast("int"))
display(ranked_avg_pitstops)

raceId name driverId forename surname position avg_pit_time 1000 Hungarian Grand Prix 838 Stoffel Vandoorne \N 21732.0 1000 Hungarian Grand Prix 1 Lewis Hamilton 1 21480.0 1000 Hungarian Grand Prix 20 Sebastian Vettel 2 23111.0 1000 Hungarian Grand Prix 8 Kimi Räikkönen 3 23150.0 1000 Hungarian Grand Prix 817 Daniel Ricciardo 4 21364.0 1000 Hungarian Grand Prix 822 Valtteri Bottas 5 21337.0 1000 Hungarian Grand Prix 842 Pierre Gasly 6 21684.0 1000 Hungarian Grand Prix 825 Kevin Magnussen 7 25126.0 1000 Hungarian Grand Prix 4 Fernando Alonso 8 21795.0 1000 Hungarian Grand Prix 832 Carlos Sainz 9 21914.0 1000 Hungarian Grand Prix 154 Romain Grosjean 10 21733.0 1000 Hungarian Grand Prix 843 Brendon Hartley 11 21831.0 1000 Hungarian Grand Prix 807 Nico Hülkenberg 12 22308.0 1000 Hungarian Grand Prix 839 Esteban Ocon 13 22258.0 1000 Hungarian Grand Prix 815 Sergio Pérez 14 22561.0 1000 Hungarian Grand Prix 828 Marcus Ericsson 15 22640.0 1000 Hungarian Grand Prix 845 Sergey Sirotkin 16 21509.0 1000 Hungarian Grand Prix 840 Lance Stroll 17 21291.0 1001 Belgian Grand Prix 8 Kimi Räikkönen \N 30909.0 1001 Belgian Grand Prix 20 Sebastian Vettel 1 20917.5 1001 Belgian Grand Prix 1 Lewis Hamilton 2 20873.5 1001 Belgian Grand Prix 830 Max Verstappen 3 21141.5 1001 Belgian Grand Prix 822 Valtteri Bottas 4 24422.333333333332 1001 Belgian Grand Prix 815 Sergio Pérez 5 21412.0 1001 Belgian Grand Prix 839 Esteban Ocon 6 21403.5 1001 Belgian Grand Prix 154 Romain Grosjean 7 21764.0 1001 Belgian Grand Prix 825 Kevin Magnussen 8 22162.5 1001 Belgian Grand Prix 842 Pierre Gasly 9 21544.5 1001 Belgian Grand Prix 828 Marcus Ericsson 10 21134.5 1001 Belgian Grand Prix 832 Carlos Sainz 11 21183.0 1001 Belgian Grand Prix 845 Sergey Sirotkin 12 21065.0 1001 Belgian Grand Prix 840 Lance Stroll 13 21316.5 1001 Belgian Grand Prix 843 Brendon Hartley 14 23700.0 1001 Belgian Grand Prix 838 Stoffel Vandoorne 15 21980.666666666668 1002 Italian Grand Prix 817 Daniel Ricciardo \N 24805.0 1002 Italian Grand Prix 154 Romain Grosjean \N 24779.0 1002 Italian Grand Prix 1 Lewis Hamilton 1 23728.0 1002 Italian Grand Prix 8 Kimi Räikkönen 2 23862.0 1002 Italian Grand Prix 822 Valtteri Bottas 3 23763.0 1002 Italian Grand Prix 20 Sebastian Vettel 4 28260.0 1002 Italian Grand Prix 830 Max Verstappen 5 24003.0 1002 Italian Grand Prix 839 Esteban Ocon 6 24827.0 1002 Italian Grand Prix 815 Sergio Pérez 7 25655.0 1002 Italian Grand Prix 832 Carlos Sainz 8 25774.0 1002 Italian Grand Prix 840 Lance Stroll 9 26440.0 1002 Italian Grand Prix 845 Sergey Sirotkin 10 25245.0 1002 Italian Grand Prix 844 Charles Leclerc 11 24084.0 1002 Italian Grand Prix 838 Stoffel Vandoorne 12 24094.0 1002 Italian Grand Prix 807 Nico Hülkenberg 13 24845.0 1002 Italian Grand Prix 842 Pierre Gasly 14 27792.0 1002 Italian Grand Prix 828 Marcus Ericsson 15 27783.5 1002 Italian Grand Prix 825 Kevin Magnussen 16 38563.0 1003 Singapore Grand Prix 1 Lewis Hamilton 1 28946.0 1003 Singapore Grand Prix 830 Max Verstappen 2 29681.0 1003 Singapore Grand Prix 20 Sebastian Vettel 3 28722.0 1003 Singapore Grand Prix 822 Valtteri Bottas 4 28542.0 1003 Singapore Grand Prix 8 Kimi Räikkönen 5 29513.0 1003 Singapore Grand Prix 817 Daniel Ricciardo 6 28779.0 1003 Singapore Grand Prix 4 Fernando Alonso 7 29922.0 1003 Singapore Grand Prix 832 Carlos Sainz 8 30242.0 1003 Singapore Grand Prix 844 Charles Leclerc 9 29310.0 1003 Singapore Grand Prix 807 Nico Hülkenberg 10 31183.0 1003 Singapore Grand Prix 828 Marcus Ericsson 11 28860.0 1003 Singapore Grand Prix 838 Stoffel Vandoorne 12 30234.0 1003 Singapore Grand Prix 842 Pierre Gasly 13 28931.0 1003 Singapore Grand Prix 840 Lance Stroll 14 28767.0 1003 Singapore Grand Prix 154 Romain Grosjean 15 29517.0 1003 Singapore Grand Prix 815 Sergio Pérez 16 28293.0 1003 Singapore Grand Prix 843 Brendon Hartley 17 29688.0 1003 Singapore Grand Prix 825 Kevin Magnussen 18 29814.333333333332 1003 Singapore Grand Prix 845 Sergey Sirotkin 19 32249.5 1004 Russian Grand Prix 843 Brendon Hartl

3. [20 pts] Insert the missing code (e.g: ALO for Alonso) for drivers based on the 'drivers' dataset

In [0]:
# If the 'code' is missing, I take the first three letters of the driver's surname
# Example: Alonso → "ALO", Schumacher → "SCH"
# This approach mimics the format of real F1 codes and ensures every driver has a non-null value.
driver_code_filled = df_drivers.withColumn(
    "code", when(col("code").isNull(), col("surname").substr(1,3)).otherwise(col("code")))
display(driver_code_filled)

driverId driverRef number code forename surname dob nationality url 1 hamilton 44 HAM Lewis Hamilton 1985-01-07 British http://en.wikipedia.org/wiki/Lewis_Hamilton 2 heidfeld \N HEI Nick Heidfeld 1977-05-10 German http://en.wikipedia.org/wiki/Nick_Heidfeld 3 rosberg 6 ROS Nico Rosberg 1985-06-27 German http://en.wikipedia.org/wiki/Nico_Rosberg 4 alonso 14 ALO Fernando Alonso 1981-07-29 Spanish http://en.wikipedia.org/wiki/Fernando_Alonso 5 kovalainen \N KOV Heikki Kovalainen 1981-10-19 Finnish http://en.wikipedia.org/wiki/Heikki_Kovalainen 6 nakajima \N NAK Kazuki Nakajima 1985-01-11 Japanese http://en.wikipedia.org/wiki/Kazuki_Nakajima 7 bourdais \N BOU Sébastien Bourdais 1979-02-28 French http://en.wikipedia.org/wiki/S%C3%A9bastien_Bourdais 8 raikkonen 7 RAI Kimi Räikkönen 1979-10-17 Finnish http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen 9 kubica 88 KUB Robert Kubica 1984-12-07 Polish http://en.wikipedia.org/wiki/Robert_Kubica 10 glock \N GLO Timo Glock 1982-03-18 German http://en.wikipedia.org/wiki/Timo_Glock 11 sato \N SAT Takuma Sato 1977-01-28 Japanese http://en.wikipedia.org/wiki/Takuma_Sato 12 piquet_jr \N PIQ Nelson Piquet Jr. 1985-07-25 Brazilian http://en.wikipedia.org/wiki/Nelson_Piquet,_Jr. 13 massa 19 MAS Felipe Massa 1981-04-25 Brazilian http://en.wikipedia.org/wiki/Felipe_Massa 14 coulthard \N COU David Coulthard 1971-03-27 British http://en.wikipedia.org/wiki/David_Coulthard 15 trulli \N TRU Jarno Trulli 1974-07-13 Italian http://en.wikipedia.org/wiki/Jarno_Trulli 16 sutil 99 SUT Adrian Sutil 1983-01-11 German http://en.wikipedia.org/wiki/Adrian_Sutil 17 webber \N WEB Mark Webber 1976-08-27 Australian http://en.wikipedia.org/wiki/Mark_Webber_(racing_driver) 18 button 22 BUT Jenson Button 1980-01-19 British http://en.wikipedia.org/wiki/Jenson_Button 19 davidson \N DAV Anthony Davidson 1979-04-18 British http://en.wikipedia.org/wiki/Anthony_Davidson 20 vettel 5 VET Sebastian Vettel 1987-07-03 German http://en.wikipedia.org/wiki/Sebastian_Vettel 21 fisichella \N FIS Giancarlo Fisichella 1973-01-14 Italian http://en.wikipedia.org/wiki/Giancarlo_Fisichella 22 barrichello \N BAR Rubens Barrichello 1972-05-23 Brazilian http://en.wikipedia.org/wiki/Rubens_Barrichello 23 ralf_schumacher \N SCH Ralf Schumacher 1975-06-30 German http://en.wikipedia.org/wiki/Ralf_Schumacher 24 liuzzi \N LIU Vitantonio Liuzzi 1980-08-06 Italian http://en.wikipedia.org/wiki/Vitantonio_Liuzzi 25 wurz \N WUR Alexander Wurz 1974-02-15 Austrian http://en.wikipedia.org/wiki/Alexander_Wurz 26 speed \N SPE Scott Speed 1983-01-24 American http://en.wikipedia.org/wiki/Scott_Speed 27 albers \N ALB Christijan Albers 1979-04-16 Dutch http://en.wikipedia.org/wiki/Christijan_Albers 28 markus_winkelhock \N WIN Markus Winkelhock 1980-06-13 German http://en.wikipedia.org/wiki/Markus_Winkelhock 29 yamamoto \N YAM Sakon Yamamoto 1982-07-09 Japanese http://en.wikipedia.org/wiki/Sakon_Yamamoto 30 michael_schumacher \N MSC Michael Schumacher 1969-01-03 German http://en.wikipedia.org/wiki/Michael_Schumacher 31 montoya \N MON Juan Pablo Montoya 1975-09-20 Colombian http://en.wikipedia.org/wiki/Juan_Pablo_Montoya 32 klien \N KLI Christian Klien 1983-02-07 Austrian http://en.wikipedia.org/wiki/Christian_Klien 33 monteiro \N TMO Tiago Monteiro 1976-07-24 Portuguese http://en.wikipedia.org/wiki/Tiago_Monteiro 34 ide \N IDE Yuji Ide 1975-01-21 Japanese http://en.wikipedia.org/wiki/Yuji_Ide 35 villeneuve \N VIL Jacques Villeneuve 1971-04-09 Canadian http://en.wikipedia.org/wiki/Jacques_Villeneuve 36 montagny \N FMO Franck Montagny 1978-01-05 French http://en.wikipedia.org/wiki/Franck_Montagny 37 rosa \N DLR Pedro de la Rosa 1971-02-24 Spanish http://en.wikipedia.org/wiki/Pedro_de_la_Rosa 38 doornbos \N DOO Robert Doornbos 1981-09-23 Dutch http://en.wikipedia.org/wiki/Robert_Doornbos 39 karthikeyan \N KAR Narain Karthikeyan 1977-01-14 Indian http://en.wikipedia.org/wiki/Narain_Karthikeyan 40 friesacher \N FRI Patrick Friesacher 1980-09-26 Austrian http://en.wikipe

4. [20 pts] Who is the youngest and oldest driver for each race? Create a new column called “Age”

In [0]:
# Join drivers and races to compute Age at race date
# Age = number of full years (birthdays) the driver has had by race day
# use floor(), datediff(race_date, dob) / 365 to computes the total days lived, 
# converts it to years, and rounds down to the most recent full year.
df_with_age = df_results.join(df_drivers, "driverId", "inner") \
    .join(df_races.select("raceId", "date"), "raceId", "left") \
    .withColumn("Age", floor(datediff(col("date"), col("dob")) / 365))

# Get min and max Age per race
age_extremes = df_with_age.groupBy("raceId").agg(
    min("Age").alias("youngest_age"),
    max("Age").alias("oldest_age")
)
# Join back to find youngest drivers
youngest = df_with_age.join(age_extremes.withColumnRenamed("raceId", "raceId_ext"), 
                            (df_with_age.raceId == col("raceId_ext")) & 
                            (df_with_age.Age == col("youngest_age")), "inner") \
                      .select(df_with_age["raceId"], "driverId", "forename", "surname", "Age") \
                      .distinct()
# Join back to find oldest drivers
oldest = df_with_age.join(age_extremes.withColumnRenamed("raceId", "raceId_ext"), 
                          (df_with_age.raceId == col("raceId_ext")) & 
                          (df_with_age.Age == col("oldest_age")), "inner") \
                    .select(df_with_age["raceId"], "driverId", "forename", "surname", "Age") \
                    .distinct()
# Display results - youngest and oldest drivers
display(youngest.orderBy("raceId"))
display(oldest.orderBy("raceId"))

raceId driverId forename surname Age 1 67 Sébastien Buemi 20 10 153 Jaime Alguersuari 19 100 32 Christian Klien 21 1000 840 Lance Stroll 19 1001 840 Lance Stroll 19 1002 840 Lance Stroll 19 1003 840 Lance Stroll 19 1004 840 Lance Stroll 19 1005 840 Lance Stroll 19 1006 840 Lance Stroll 19 1007 840 Lance Stroll 20 1008 840 Lance Stroll 20 1009 840 Lance Stroll 20 101 32 Christian Klien 21 1010 846 Lando Norris 19 1011 846 Lando Norris 19 1012 846 Lando Norris 19 1013 846 Lando Norris 19 1014 846 Lando Norris 19 1015 846 Lando Norris 19 1016 846 Lando Norris 19 1017 846 Lando Norris 19 1018 846 Lando Norris 19 1019 846 Lando Norris 19 102 32 Christian Klien 21 1020 846 Lando Norris 19 1021 846 Lando Norris 19 1022 846 Lando Norris 19 1023 846 Lando Norris 19 1024 846 Lando Norris 19 1025 846 Lando Norris 19 1026 846 Lando Norris 19 1027 846 Lando Norris 19 1028 846 Lando Norris 19 1029 846 Lando Norris 20 103 32 Christian Klien 21 1030 846 Lando Norris 20 1031 846 Lando Norris 20 1032 846 Lando Norris 20 1033 846 Lando Norris 20 1034 846 Lando Norris 20 1035 846 Lando Norris 20 1036 846 Lando Norris 20 1037 846 Lando Norris 20 1038 846 Lando Norris 20 1039 846 Lando Norris 20 104 32 Christian Klien 21 1040 846 Lando Norris 20 1041 846 Lando Norris 20 1042 846 Lando Norris 20 1043 846 Lando Norris 20 1044 846 Lando Norris 21 1045 846 Lando Norris 21 1046 846 Lando Norris 21 1047 846 Lando Norris 21 105 32 Christian Klien 21 1051 852 Yuki Tsunoda 21 1052 852 Yuki Tsunoda 20 1053 852 Yuki Tsunoda 20 1054 852 Yuki Tsunoda 20 1055 846 Lando Norris 21 1055 852 Yuki Tsunoda 21 1056 846 Lando Norris 21 1056 852 Yuki Tsunoda 21 1057 846 Lando Norris 21 1057 852 Yuki Tsunoda 21 1058 846 Lando Norris 21 1058 852 Yuki Tsunoda 21 1059 846 Lando Norris 21 1059 852 Yuki Tsunoda 21 106 32 Christian Klien 21 1060 846 Lando Norris 21 1060 852 Yuki Tsunoda 21 1061 846 Lando Norris 21 1061 852 Yuki Tsunoda 21 1062 852 Yuki Tsunoda 21 1062 846 Lando Norris 21 1063 846 Lando Norris 21 1063 852 Yuki Tsunoda 21 1064 846 Lando Norris 21 1064 852 Yuki Tsunoda 21 1065 846 Lando Norris 21 1065 852 Yuki Tsunoda 21 1066 846 Lando Norris 21 1066 852 Yuki Tsunoda 21 1067 846 Lando Norris 21 1067 852 Yuki Tsunoda 21 1069 846 Lando Norris 21 1069 852 Yuki Tsunoda 21 107 32 Christian Klien 21 1070 852 Yuki Tsunoda 21 1071 852 Yuki Tsunoda 21 1072 852 Yuki Tsunoda 21 1073 852 Yuki Tsunoda 21 1074 852 Yuki Tsunoda 21 1075 852 Yuki Tsunoda 21 1076 852 Yuki Tsunoda 21 1077 852 Yuki Tsunoda 21 1078 852 Yuki Tsunoda 22 1078 846 Lando Norris 22 1078 855 Guanyu Zhou 22 1079 846 Lando Norris 22 1079 852 Yuki Tsunoda 22 1079 855 Guanyu Zhou 22 108 4 Fernando Alonso 21 1080 846 Lando Norris 22 1080 852 Yuki Tsunoda 22 1081 846 Lando Norris 22 1081 852 Yuki Tsunoda 22 1082 846 Lando Norris 22 1082 852 Yuki Tsunoda 22 1083 846 Lando Norris 22 1083 852 Yuki Tsunoda 22 1084 846 Lando Norris 22 1084 852 Yuki Tsunoda 22 1085 846 Lando Norris 22 1085 852 Yuki Tsunoda 22 1086 846 Lando Norris 22 1086 852 Yuki Tsunoda 22 1087 846 Lando Norris 22 1087 852 Yuki Tsunoda 22 1088 846 Lando Norris 22 1088 852 Yuki Tsunoda 22 1089 846 Lando Norris 22 1089 852 Yuki Tsunoda 22 109 4 Fernando Alonso 21 1091 846 Lando Norris 22 1091 852 Yuki Tsunoda 22 1092 846 Lando Norris 22 1092 852 Yuki Tsunoda 22 1093 846 Lando Norris 22 1093 852 Yuki Tsunoda 22 1094 846 Lando Norris 22 1094 852 Yuki Tsunoda 22 1095 852 Yuki Tsunoda 22 1096 852 Yuki Tsunoda 22 1098 857 Oscar Piastri 21 1099 857 Oscar Piastri 21 11 153 Jaime Alguersuari 19 110 4 Fernando Alonso 21 1100 857 Oscar Piastri 22 1100 852 Yuki Tsunoda 22 1100 858 Logan Sargeant 22 1101 852 Yuki Tsunoda 22 1101 857 Oscar Piastri 22 1101 858 Logan Sargeant 22 1102 857 Oscar Piastri 22 1102 858 Logan Sargeant 22 1104 857 Oscar Piastri 22 1104 858 Logan Sargeant 22 1105 857 Oscar Piastri 22 1105 858 Logan Sargeant 22 1106 857 Oscar Piastri 22 1106 858 Logan Sargeant 22 1107 858 Logan Sargeant 22 1107 857 Oscar Piastri 22 1108 857 Oscar Piastri 22 110

raceId driverId forename surname Age 1 22 Rubens Barrichello 36 1 21 Giancarlo Fisichella 36 10 22 Rubens Barrichello 37 100 44 Olivier Panis 37 1000 8 Kimi Räikkönen 38 1001 8 Kimi Räikkönen 38 1002 8 Kimi Räikkönen 38 1003 8 Kimi Räikkönen 38 1004 8 Kimi Räikkönen 38 1005 8 Kimi Räikkönen 39 1006 8 Kimi Räikkönen 39 1007 8 Kimi Räikkönen 39 1008 8 Kimi Räikkönen 39 1009 8 Kimi Räikkönen 39 101 44 Olivier Panis 37 1010 8 Kimi Räikkönen 39 1011 8 Kimi Räikkönen 39 1012 8 Kimi Räikkönen 39 1013 8 Kimi Räikkönen 39 1014 8 Kimi Räikkönen 39 1015 8 Kimi Räikkönen 39 1016 8 Kimi Räikkönen 39 1017 8 Kimi Räikkönen 39 1018 8 Kimi Räikkönen 39 1019 8 Kimi Räikkönen 39 102 44 Olivier Panis 37 1020 8 Kimi Räikkönen 39 1021 8 Kimi Räikkönen 39 1022 8 Kimi Räikkönen 39 1023 8 Kimi Räikkönen 39 1024 8 Kimi Räikkönen 39 1025 8 Kimi Räikkönen 39 1026 8 Kimi Räikkönen 40 1027 8 Kimi Räikkönen 40 1028 8 Kimi Räikkönen 40 1029 8 Kimi Räikkönen 40 103 44 Olivier Panis 38 1030 8 Kimi Räikkönen 40 1031 8 Kimi Räikkönen 40 1032 8 Kimi Räikkönen 40 1033 8 Kimi Räikkönen 40 1034 8 Kimi Räikkönen 40 1035 8 Kimi Räikkönen 40 1036 8 Kimi Räikkönen 40 1037 8 Kimi Räikkönen 40 1038 8 Kimi Räikkönen 40 1039 8 Kimi Räikkönen 40 104 44 Olivier Panis 38 1040 8 Kimi Räikkönen 40 1041 8 Kimi Räikkönen 41 1042 8 Kimi Räikkönen 41 1043 8 Kimi Räikkönen 41 1044 8 Kimi Räikkönen 41 1045 8 Kimi Räikkönen 41 1046 8 Kimi Räikkönen 41 1047 8 Kimi Räikkönen 41 105 44 Olivier Panis 38 1051 8 Kimi Räikkönen 42 1052 8 Kimi Räikkönen 41 1053 8 Kimi Räikkönen 41 1054 8 Kimi Räikkönen 41 1055 8 Kimi Räikkönen 41 1056 8 Kimi Räikkönen 41 1057 8 Kimi Räikkönen 41 1058 8 Kimi Räikkönen 41 1059 8 Kimi Räikkönen 41 106 44 Olivier Panis 38 1060 8 Kimi Räikkönen 41 1061 8 Kimi Räikkönen 41 1062 8 Kimi Räikkönen 41 1063 8 Kimi Räikkönen 41 1064 4 Fernando Alonso 40 1065 4 Fernando Alonso 40 1066 8 Kimi Räikkönen 41 1067 8 Kimi Räikkönen 42 1069 8 Kimi Räikkönen 42 107 30 Michael Schumacher 35 1070 8 Kimi Räikkönen 42 1071 8 Kimi Räikkönen 42 1072 8 Kimi Räikkönen 42 1073 8 Kimi Räikkönen 42 1074 4 Fernando Alonso 40 1075 4 Fernando Alonso 40 1076 4 Fernando Alonso 40 1077 4 Fernando Alonso 40 1078 4 Fernando Alonso 40 1079 4 Fernando Alonso 40 108 44 Olivier Panis 36 1080 4 Fernando Alonso 40 1081 4 Fernando Alonso 40 1082 4 Fernando Alonso 40 1083 4 Fernando Alonso 40 1084 4 Fernando Alonso 40 1085 4 Fernando Alonso 41 1086 4 Fernando Alonso 41 1087 4 Fernando Alonso 41 1088 4 Fernando Alonso 41 1089 4 Fernando Alonso 41 109 44 Olivier Panis 36 1091 4 Fernando Alonso 41 1092 4 Fernando Alonso 41 1093 4 Fernando Alonso 41 1094 4 Fernando Alonso 41 1095 4 Fernando Alonso 41 1096 4 Fernando Alonso 41 1098 4 Fernando Alonso 41 1099 4 Fernando Alonso 41 11 69 Luca Badoer 38 110 44 Olivier Panis 36 1100 4 Fernando Alonso 41 1101 4 Fernando Alonso 41 1102 4 Fernando Alonso 41 1104 4 Fernando Alonso 41 1105 4 Fernando Alonso 41 1106 4 Fernando Alonso 41 1107 4 Fernando Alonso 41 1108 4 Fernando Alonso 41 1109 4 Fernando Alonso 42 111 44 Olivier Panis 36 1110 4 Fernando Alonso 42 112 44 Olivier Panis 36 113 44 Olivier Panis 36 113 49 Heinz-Harald Frentzen 36 114 44 Olivier Panis 36 114 49 Heinz-Harald Frentzen 36 115 44 Olivier Panis 36 115 49 Heinz-Harald Frentzen 36 116 49 Heinz-Harald Frentzen 36 116 44 Olivier Panis 36 117 44 Olivier Panis 36 117 49 Heinz-Harald Frentzen 36 118 44 Olivier Panis 36 118 49 Heinz-Harald Frentzen 36 119 44 Olivier Panis 36 119 49 Heinz-Harald Frentzen 36 12 69 Luca Badoer 38 120 44 Olivier Panis 37 121 44 Olivier Panis 37 122 44 Olivier Panis 37 123 44 Olivier Panis 37 124 56 Eddie Irvine 36 125 56 Eddie Irvine 36 126 56 Eddie Irvine 36 127 56 Eddie Irvine 36 128 56 Eddie Irvine 36 129 56 Eddie Irvine 36 13 22 Rubens Barrichello 37 130 56 Eddie Irvine 36 131 56 Eddie Irvine 36 132 56 Eddie Irvine 36 133 56 Eddie Irvine 36 134 56 Eddie Irvine 36 135 56 Eddie Irvine 36 136 56 Eddie Irvine 36 137 56 Eddie Irvine 36 137 44 Olivier Panis 36 138 56 Eddie Irvine 36

5. [20 pts] For a given race, which driver has the most wins and losses?

In [0]:
# Join race dates into results
results_with_dates = df_results.join(df_races.select("raceId", "date"), on="raceId", how="inner") \
                               .select("raceId", "driverId", "positionOrder", "date")

# Define time-aware window per driver by using a window specification
window_spec = Window.partitionBy("driverId").orderBy("date").rowsBetween(Window.unboundedPreceding, Window.currentRow)

# Calculate cumulative wins and losses up to each race
driver_stats = results_with_dates.withColumn(
    "Cumulative_Wins",
    F.sum(F.when(F.col("positionOrder") == 1, 1).otherwise(0)).over(window_spec)
).withColumn(
    "Cumulative_Losses",
    F.sum(F.when((F.col("positionOrder") != 1) & (F.col("positionOrder").isNotNull()), 1).otherwise(0)).over(window_spec)
)
# Get the driver with the most cumulative wins per race
driver_stats_alias = driver_stats.alias("ds")
most_wins_per_race = driver_stats.groupBy("raceId").agg(
    F.max("Cumulative_Wins").alias("Max_Wins")
).alias("mw").join(driver_stats_alias,
                   (driver_stats_alias["Cumulative_Wins"] == F.col("mw.Max_Wins")) & 
                   (driver_stats_alias["raceId"] == F.col("mw.raceId"))
).select("ds.raceId", "ds.driverId", "ds.Cumulative_Wins", "ds.Cumulative_Losses")
# Get the driver with the most cumulative losses per race
most_losses_per_race = driver_stats.groupBy("raceId").agg(
    F.max("Cumulative_Losses").alias("Max_Losses")
).alias("ml").join(driver_stats_alias,
                   (driver_stats_alias["Cumulative_Losses"] == F.col("ml.Max_Losses")) & 
                   (driver_stats_alias["raceId"] == F.col("ml.raceId"))
).select("ds.raceId", "ds.driverId", "ds.Cumulative_Wins", "ds.Cumulative_Losses")
# Display results - most wins and losses per race
display(most_wins_per_race.orderBy("raceId"))
display(most_losses_per_race.orderBy("raceId"))

raceId driverId Cumulative_Wins Cumulative_Losses 1 4 21 103 10 4 21 112 100 30 80 126 1000 1 67 153 1001 1 67 154 1002 1 68 154 1003 1 69 154 1004 1 70 154 1005 1 71 154 1006 1 71 155 1007 1 71 156 1008 1 72 156 1009 1 73 156 101 30 81 126 1010 1 73 157 1011 1 74 157 1012 1 75 157 1013 1 75 158 1014 1 76 158 1015 1 77 158 1016 1 78 158 1017 1 79 158 1018 1 79 159 1019 1 80 159 102 30 82 126 1020 1 80 160 1021 1 81 160 1022 1 81 161 1023 1 81 162 1024 1 81 163 1025 1 82 163 1026 1 82 164 1027 1 83 164 1028 1 83 165 1029 1 83 166 103 30 82 127 1030 1 84 166 1031 1 84 167 1032 1 85 167 1033 1 86 167 1034 1 87 167 1035 1 87 168 1036 1 88 168 1037 1 89 168 1038 1 89 169 1039 1 90 169 104 30 82 128 1040 1 90 170 1041 1 91 170 1042 1 92 170 1043 1 93 170 1044 1 94 170 1045 1 95 170 1046 20 53 204 1047 1 95 171 105 30 82 129 1051 1 102 184 1052 1 96 171 1053 1 96 172 1054 1 97 172 1055 1 98 172 1056 1 98 173 1057 1 98 174 1058 1 98 176 1059 1 98 175 106 30 83 129 1060 1 98 177 1061 1 99 177 1062 1 99 178 1063 1 99 179 1064 1 99 180 1065 1 99 181 1066 1 100 181 1067 1 100 182 1069 1 100 183 107 30 83 130 1070 1 100 184 1071 1 101 184 1072 1 103 184 1073 1 103 185 1074 1 103 186 1075 1 103 187 1076 1 103 188 1077 1 103 189 1078 1 103 190 1079 1 103 191 108 30 64 116 1080 1 103 192 1081 1 103 193 1082 1 103 194 1083 1 103 195 1084 1 103 196 1085 1 103 197 1086 1 103 198 1087 1 103 199 1088 1 103 200 1089 1 103 201 109 30 64 117 1091 1 103 202 1092 1 103 203 1093 1 103 204 1094 1 103 205 1095 1 103 206 1096 1 103 207 1098 1 103 208 1099 1 103 209 11 4 21 113 110 30 64 118 1100 1 103 210 1101 1 103 211 1102 1 103 212 1104 1 103 213 1105 1 103 214 1106 1 103 215 1107 1 103 216 1108 1 103 217 1109 1 103 218 111 30 65 118 1110 1 103 219 112 30 66 118 113 30 67 118 114 30 67 119 115 30 68 119 116 30 68 120 117 30 68 121 118 30 68 122 119 30 68 123 12 4 21 114 120 30 68 124 121 30 69 124 122 30 70 124 123 30 70 125 124 30 54 109 125 30 54 110 126 30 55 110 127 30 56 110 128 30 57 110 129 30 58 110 13 4 21 115 130 30 58 111 131 30 59 111 132 30 59 112 133 30 60 112 134 30 61 112 135 30 62 112 136 30 62 113 137 30 63 113 138 30 63 114 139 30 63 115 14 4 21 116 140 30 64 115 141 30 45 101 142 30 46 101 143 30 46 102 144 30 46 103 145 30 47 103 146 30 47 104 147 30 48 104 148 30 48 105 149 30 49 105 15 4 21 117 150 30 50 105 151 30 50 106 152 30 50 107 153 30 51 107 154 30 52 107 155 30 52 108 156 30 52 109 157 30 53 109 158 30 36 93 159 30 37 93 16 4 21 118 160 30 38 93 161 30 38 94 162 30 38 95 163 30 39 95 164 30 39 96 165 30 40 96 166 30 40 97 167 30 40 98 168 30 40 99 169 30 40 100 17 4 21 119 170 30 40 101 171 30 41 101 172 30 42 101 173 30 43 101 174 30 44 101 175 30 33 86 176 30 33 87 177 30 34 87 178 30 35 87 179 30 35 88 18 4 19 87 180 30 35 89 181 30 35 90 182 30 35 91 183 71 22 93 184 71 22 94 185 71 22 95 186 71 22 96 187 71 22 97 188 71 22 98 189 30 35 92 19 4 19 88 190 30 35 93 191 30 27 76 192 30 27 77 193 30 28 77 194 30 28 78 195 30 28 79 196 30 28 80 197 30 29 80 198 30 30 80 199 30 31 80 2 4 21 104 20 4 19 89 200 30 31 81 201 30 31 82 202 30 32 82 203 30 32 83 204 30 33 83 205 30 33 84 206 30 33 85 207 30 22 64 208 30 22 65 209 30 22 66 21 4 19 90 210 30 22 67 211 30 23 67 212 30 23 68 213 30 24 68 214 30 25 68 215 30 25 69 216 30 25 70 217 30 25 71 218 30 26 71 219 30 26 72 22 4 19 91 220 30 26 73 221 30 26 74 222 30 27 74 223 30 27 75 224 30 19 51 225 30 19 52 226 30 19 53 227 30 19 54 228 30 19 55 229 30 19 56 23 4 19 92 230 30 20 56 231 30 20 57 232 30 20 58 233 30 20 59 234 30 20 60 234 71 20 48 235 30 20 61 235 71 20 49 236 30 21 61 237 30 22 61 238 30 22 62 239 30 22 63 24 4 19 93 240 30 11 42 241 30 11 43 242 95 31 160 243 95 31 161 244 30 13 44 245 30 13 45 246 30 14 45 247 30 14 46 248 30 15 46 249 30 15 47 25 4 19 94 250 30 16 47 251 30 16 48 252 30 16 49 253 30 17 49 254 30 18 49 255 30 19 49 256 30 19 50 257 102 41 119 258 102 41 120 259 102 41 121 26 4 19 95 260 77 8 143 261 77 8 144 262 77 8 145 263 95 30 157 264

raceId driverId Cumulative_Wins Cumulative_Losses 1 22 9 263 10 22 9 272 100 22 7 185 1000 4 32 273 1001 4 32 274 1002 4 32 275 1003 4 32 276 1004 4 32 277 1005 4 32 278 1006 4 32 279 1007 4 32 280 1008 4 32 281 1009 4 32 282 101 22 7 186 1010 8 21 274 1011 8 21 275 1012 8 21 276 1013 8 21 277 1014 8 21 278 1015 8 21 279 1016 8 21 280 1017 8 21 281 1018 8 21 282 1019 8 21 283 102 22 7 187 1020 8 21 284 1021 8 21 285 1022 8 21 286 1023 8 21 287 1024 8 21 288 1025 8 21 289 1026 8 21 290 1027 8 21 291 1028 8 21 292 1029 8 21 293 103 22 7 188 1030 8 21 294 1031 8 21 295 1032 8 21 296 1033 8 21 297 1034 8 21 298 1035 8 21 299 1036 8 21 300 1037 8 21 301 1038 8 21 302 1039 8 21 303 104 22 8 188 1040 8 21 304 1041 8 21 305 1042 8 21 306 1043 8 21 307 1044 8 21 308 1045 8 21 309 1046 8 21 310 1047 8 21 311 105 22 9 188 1051 8 21 329 1052 8 21 312 1053 8 21 313 1054 8 21 314 1055 8 21 315 1056 8 21 316 1057 8 21 317 1058 8 21 319 1059 8 21 318 106 22 9 189 1060 8 21 320 1061 8 21 321 1062 8 21 322 1063 8 21 323 1064 4 32 295 1065 4 32 296 1066 8 21 324 1067 8 21 325 1069 8 21 326 107 22 9 190 1070 8 21 327 1071 8 21 328 1072 8 21 330 1073 8 21 331 1074 4 32 305 1075 4 32 306 1076 4 32 307 1077 4 32 308 1078 4 32 309 1079 4 32 310 108 22 5 161 1080 4 32 311 1081 4 32 312 1082 4 32 313 1083 4 32 314 1084 4 32 315 1085 4 32 316 1086 4 32 317 1087 4 32 318 1088 4 32 319 1089 4 32 320 109 22 5 162 1091 4 32 321 1092 4 32 322 1093 4 32 323 1094 4 32 324 1095 4 32 325 1096 4 32 326 1098 4 32 327 1099 4 32 328 11 22 10 272 110 22 5 163 1100 4 32 329 1101 4 32 330 1102 4 32 331 1104 4 32 332 1105 4 32 333 1106 4 32 334 1107 4 32 335 1108 4 32 336 1109 4 32 337 111 22 5 164 1110 4 32 338 112 22 5 165 113 22 5 166 114 22 5 167 115 22 5 168 116 22 5 169 117 22 5 170 118 22 6 170 119 22 6 171 12 22 10 273 120 22 6 172 121 22 6 173 122 22 6 174 123 22 7 174 124 22 1 148 125 22 1 149 126 22 1 150 127 22 1 151 128 22 1 152 129 22 1 153 13 22 11 273 130 22 1 154 131 22 1 155 132 22 2 155 133 22 2 156 134 22 2 157 135 22 2 158 136 22 3 158 137 22 3 159 138 22 4 159 139 22 5 159 14 22 11 274 140 22 5 160 141 55 1 185 142 55 1 186 143 55 1 187 144 55 1 188 145 55 1 189 146 55 1 190 147 55 1 191 148 55 1 192 149 55 1 193 15 22 11 275 150 55 1 194 151 55 1 195 152 55 1 196 153 55 1 197 154 55 1 198 155 55 1 199 156 55 1 200 157 55 1 201 158 55 1 168 159 55 1 169 16 22 11 276 160 55 1 170 161 55 1 171 162 55 1 172 163 55 1 173 164 55 1 174 165 55 1 175 166 55 1 176 167 55 1 177 168 55 1 178 169 55 1 179 17 22 11 277 170 55 1 180 171 55 1 181 172 55 1 182 173 55 1 183 174 55 1 184 175 55 1 152 176 55 1 153 177 55 1 154 178 55 1 155 179 55 1 156 18 22 9 245 180 55 1 157 181 55 1 158 182 55 1 159 183 55 1 160 184 55 1 161 185 55 1 162 186 55 1 163 187 55 1 164 188 55 1 165 189 55 1 166 19 22 9 246 190 55 1 167 191 55 1 136 192 55 1 137 193 55 1 138 194 55 1 139 195 55 1 140 196 55 1 141 197 55 1 142 198 55 1 143 199 55 1 144 2 22 9 264 20 22 9 247 200 55 1 145 201 55 1 146 202 55 1 147 203 55 1 148 204 55 1 149 205 55 1 150 206 55 1 151 207 77 9 188 208 77 9 189 209 77 9 190 21 22 9 248 210 77 9 191 211 77 9 192 212 77 9 193 213 55 1 125 214 55 1 126 215 55 1 127 216 77 10 193 217 77 10 194 218 77 10 195 219 77 10 196 22 22 9 249 220 77 10 197 221 77 10 198 222 77 10 199 223 77 10 200 224 77 9 172 225 77 9 173 226 77 9 174 227 77 9 175 228 77 9 176 229 77 9 177 23 22 9 250 230 77 9 178 231 77 9 179 232 77 9 180 233 77 9 181 234 77 9 182 235 77 9 183 236 77 9 184 237 77 9 185 238 77 9 186 239 77 9 187 24 22 9 251 240 77 9 155 241 77 9 156 242 95 31 160 243 95 31 161 244 77 9 159 245 77 9 160 246 77 9 161 247 77 9 162 248 77 9 163 249 77 9 164 25 22 9 252 250 77 9 165 251 77 9 166 252 77 9 167 253 77 9 168 254 77 9 169 255 77 9 170 256 77 9 171 257 105 5 195 258 105 5 196 259 110 0 204 26 22 9 253 260 110 0 205 261 105 5 199 262 110 0 206 263 110 0 207 264 110 0 208 265 110 0 209 266 110 0 210 267 110 0 211 268 110 0 212 269 110 0 213 27 22 9 254 270 110 0 214 27

6. [10 pts] Continue exploring the data by answering your own question.

In [0]:
# My question: Which driver has completed the most races in their career?

# Filter for races the driver actually finished (has a position)
# This removes Did Not Finish and disqualified entries from the dataset by using isNonNull()
completed_races = df_results.filter(col("position").isNotNull())
# Count how many races each driver has completed by using count()
race_counts = completed_races.groupBy("driverId").agg(count("raceId").alias("races_completed"))
# Join with driver names
driver_names = df_drivers.select("driverId", "forename", "surname")
most_active_drivers = race_counts.join(driver_names, "driverId") \
    .orderBy(col("races_completed").desc())
# Display the top 10 drivers who have completed the most races
display(most_active_drivers.limit(10))

driverId races_completed forename surname 4 370 Fernando Alonso 8 352 Kimi Räikkönen 22 326 Rubens Barrichello 1 322 Lewis Hamilton 18 309 Jenson Button 30 308 Michael Schumacher 20 300 Sebastian Vettel 13 271 Felipe Massa 119 257 Riccardo Patrese 15 256 Jarno Trulli

**ANSWER:** Based on the output above, the driver who has completed the most races is **Fernando Alonso**, with a total of **370 races completed**.